### PR TITLE
Media Ranker

### PR DESCRIPTION
**Pull request by sheland**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I created custom routes for my session controller. For example post 'users/login', to: 'sessions#login', as: 'login' creates a user by having them log in into session. Because of this I did not need to create a route for Users#create. 
Describe how you approached testing that model method I first created the tests and then tried to make them pass. If it failed because I didn't add add a validation, I added it, then ran the rest again. 
. What edge cases did you come up with? | Incomplete users fields, saving nil valued variables, 
What are `session` and `flash`? What is the difference between them? | session is used to login a user, session keeps the user's data in cookies until they logout. Their data is then destroyed. Flash is a one-time message that the user gets to see after the controller successfully or successfully completes an action.  
Describe a controller filter you wrote. | It dries up your code be reducing repetitive lines of codes and saving it in a method. 
What was one thing that you gained more clarity on through this assignment? | I gained more clarity about model's relationships, how they connect to each other by creating references which then lets you use many dot.methods to retrieve interesting data. 
What is the Heroku URL of your deployed application? | 
Do you have any recommendations on how we could improve this project for the next cohort? |  perhaps more time to complete it.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  root 'works#homepage'
+
+  resources :works do
+    resources :votes, only: [:create]
+  end
+
+  resources :users,
+    only: [:index, :show]
+  end
```
Comment (dHelmgren) [inline config/routes.rb:17]: Nice work limiting the routes for users to be just the ones you need!

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        </li>
+      </ul>
+    </section>
+
+
+    <section class="flash">
+      <% flash.each do |name, message| %>
+        <% if message.class == Array %>
+        <div class="alert-warning">
+          <% message.each do |msg| %>
```
Comment (dHelmgren) [inline app/views/layouts/application.html.erb:55]: I might move this flash block outside of the nav. 1) it doesn't actually affect navigation, but 2) it's placement is not easy to read.

File: app/views/works/homepage.html.erb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  %>
+  <%= render partial: "top_ten", locals: { #render top books from _top_ten.html.erb
+    work_category: "book",
+    data: @books
+  }
+  %>
+  <%= render partial: "top_ten", locals: { #render top albums from _top_ten.html.erb
+    work_category: "album",
+    data: @albums
+  }
```
Comment (dHelmgren) [inline app/views/works/homepage.html.erb:29]: Nice job using view partials here, but can you think of a way to further DRY this code?

File: app/views/works/homepage.html.erb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+<% if @best_work %>
+  <section class="spotlight">
+    <h2 class="spotlight__header">
+      <span class="spotlight__header--prefix">Media Spotlight</span>
+      <%= link_to @best_work.title, work_path(@best_work), class: "spotlight__link-to" %> by <%= @best_work.creator %>
```
Comment (dHelmgren) [inline app/views/works/homepage.html.erb:5]: You don't define a `@best_work`, so this ERB doesn't run. The place to do this would be in the controller that directs the user to the main page.

File: app/models/vote.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
```
Comment (dHelmgren) [inline app/models/vote.rb:3]: You can't delete a work that has votes! This is because you've created a link between the two tables at the database level, and deleting the work would leave behind invalid votes. You might want to check out the [dependent](http://guides.rubyonrails.org/association_basics.html#dependent) argument to `has_many`.

File: app/models/work.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    top_ten("album")  #displays up to 10 albums
+  end
+
+  def self.highest_books #call self.top_10 method for book category
+    top_ten("book") #displays up to 10 books
+  end
+
+  def self.highest_movies #call self.top_10 method for movie category
+    top_ten("movie")  #displays up to 10 movies
+  end
```
Comment (dHelmgren) [inline app/models/work.rb:52]: You have the same code repeated 3 times here. Could you use a parameter to help DRY this code?

File: app/models/work.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+
```
Comment (dHelmgren) [inline app/models/work.rb:54]: This would be a great place for a method that found the `@best_work`!

File: app/models/work.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  has_many :votes
+  has_many :users, through: :votes
+
+
+  def upvote(user)
+    @vote = Vote.new #create a new vote
+    @vote.work_id = self.id #a work's id is assigned vote's foreign key, work_id
+    @vote.user_id = user.id #a user's id is assigned to vote's foreign key, user_id
+    @vote.vote_time = Date.today #vote's date
+    unless @vote.valid? #if vote is not valid...
```
Comment (dHelmgren) [inline app/models/work.rb:13]: Either here, or in your vote model you should be validating that each vote is coming from a unique user!

File: app/models/work.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  CATEGORIES = %w(album book movie) #store categories in a constant
+
+  has_many :votes
+  has_many :users, through: :votes
+
+
+  def upvote(user)
+    @vote = Vote.new #create a new vote
+    @vote.work_id = self.id #a work's id is assigned vote's foreign key, work_id
+    @vote.user_id = user.id #a user's id is assigned to vote's foreign key, user_id
```
Comment (dHelmgren) [inline app/models/work.rb:11]: This method doesn't validate whether or not the user is valid, which means if you try to upvote without validating the user, this method breaks!

File: test/models/vote_test.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    it 'belongs to a user' do
+      # Arrange (done with let)
+
+      #Act
+      user = vote.user #a vote belonging to a user
+
+      #Assert
+      expect(vote).must_be_instance_of User
+      expect(user.id).must_equal vote.user_id
```
Comment (dHelmgren) [inline test/models/vote_test.rb:39]: I would be trying to break the uniqueness property here rather than in the tests for works. Specific test cases I'd want to see:

- A user can have votes for two different works
- A work can have votes from two different users
- A user cannot vote for the same work twice

File: test/models/work_test.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    #Arrange
+    #Act
+    user = User_id(id: 2)
+    work = self.id(id: 22)
+    @vote = Vote.id.include(user_id: 2, self.id: 22)
+
+    #Assert
+    expect(@vote).must_equal false
+    expect(work.errors.messages).wont_include: user_id
+
```
Comment (dHelmgren) [inline test/models/work_test.rb:94]: There's a lot of interesting test cases for your custom `Work` methods missing here. For your category and top ten methods, I would ask:

- What if there are no works of that category?
- What if there are less than 10 works?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

File: app/models/work.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+
+  # def self.top
+  #   works = Work.all.to_a
+  #   return works.max_by { |work| work.total_votes }
+  # end
+
+  def self.highest_albums #call self.top_10 method for album category
+    top_ten("album")  #displays up to 10 albums
```
Comment (dHelmgren) [inline app/models/work.rb:43]: You should use symbols instead of strings here because your categories don't show up properly when they aren't capitalized the same way. That is, you have "albums" here and "Albums" in your seeds, so some of your work never shows up.

File: app/models/work.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+  def upvote(user)
+    @vote = Vote.new #create a new vote
+    @vote.work_id = self.id #a work's id is assigned vote's foreign key, work_id
+    @vote.user_id = user.id #a user's id is assigned to vote's foreign key, user_id
+    @vote.vote_time = Date.today #vote's date
+    unless @vote.valid? #if vote is not valid...
+      puts "#{@vote.errors.messages}" #display error messages
+    end
+    @vote.save #save vote so user cannot duplicate vote for a work
```
Comment (dHelmgren) [inline app/models/work.rb:16]: This method should add votes to list of votes for a work. Your vote count doesn't get updated properly, so your sorts never work.

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | Good
Comprehension questions	| Good
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Views are well-organized (DRY, use of semantic HTML, use of partials) | Good, see comments for further improvements!
Errors are reported to the user | No, see comments
Business logic lives in the models | Yes
Models are thoroughly tested, including relations, validations and any custom logic | See comments
**Wave 1 - Media** |
Splash page shows the three media categories | Yes
Basic CRUD operations on media are present and functional | Yes, with caveats, see comments
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | no
A user cannot vote for the same media more than once | No, see comments
All media lists are ordered by vote count | No, see comments
Splash page contains a media spotlight | No
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | Yes
Individual user pages and the user list are present | Yes
**Optional - Styling** |
Bootstrap is used appropriately | Yes
Look and feel is similar to the original | yes
**Overall** | While I think there are a lot of things working in favor of this project, I don't think it's completely meeting the learning goals we set out. Specifically, I see there are a variety of errors around your CRUD actions that don't break cleanly, you aren't properly utilizing session via the sessions controller, and your validations aren't implemented properly and at time contribute to your errors. That said, I think there are a lot of things you did well too: your view partials are excellent, you made good use of categories, and your model methods are working correctly on their own. It all needs to come together though, and unfortunately here it didn't quite.
